[Duong Minh Duc] iP

Masunori opened the pull request:
<!-- Heading -->
# Mirai

<!-- Blockquote -->
<!-- Emoji -->
> "Come on! It's only 8 in the morning..."
> "Nope. As long as I am here, you will not revert to that sedentary lifestyle again."
> "I thought you were just a fictional character?"
> "Well, not anymore 😉"

Once a fictional artificial intelligence entity with sentience, now, your favourite heroine Mirai has bloomed to life as a personal assistant! In the current iteration, Mirai:
<!-- Bullet list -->
- Supports both a friendly graphical user interface (and command-line interface for programmers)
- Has an intuitive and robust command set, easy to learn and remember

All you need to do is,
<!-- Numbered list -->
<!-- Hyperlink -->
1. download Mirai from [here](https://github.com/Masunori/ip/releases/download/A-Jar/mirai.jar).
2. double-click on the icon
3. Mirai is ready to use!

<!-- Inline code -->
> [!TIP]
> Run `help` to see the list of commands Mirai can understand.

<!-- text-formatting (bold) -->
And it is completely **FREE**!!!

Features:
<!-- Task list -->
- [x] Managing tasks
- [ ] Managing deadlines (coming soon...)
- [ ] Reminders (coming soon...)

## For Java programmers
The `main` method is:
<!-- Fenced code block with syntax highlighting -->
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

You can freely modify the source code.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Masunori committed:
Rename and add greeting to main class.

Masunori committed:
-Customise greeting
- Add 'echo' function

Masunori committed:
Reformat the chatbot into an object that exposes two methods:
- public Mirai(): Constructor.
- public void startConversation(): initiate a conversation with the Mirai object.

Masunori committed:
Update behaviour for Mirai:
- bye: Goodbye the user and terminate the conversation.
- list: List all text in the conversation, except the specified commands (bye, list)
- default: Store the text.

Masunori committed:
- Added a Task class to abstract a task that is done/undone.
- Added a Command interface to abstract a user command from the CLI.
- Replace the if/else for command using a HashMap that maps from user command (String) to the action (Command).
- A Command takes in a varargs of type String, and return a boolean, a signal to whether to terminate the conversation.
- Added two tasks: markTask for the 'mark' command and unmarkTask for the 'unmark' command.

Masunori committed:
Add three subclasses for Task class:
- ToDo: encapsulates a Task with no date/time attached.
- Deadline: encapsulates a Task with a deadline as a date/time.
- Event: encapsulates a Task with start date/time and end date/time.

Masunori committed:
Add three subclasses for Task class:
- ToDo: encapsulates a Task with no date/time attached.
- Deadline: encapsulates a Task with a deadline as a date/time.
- Event: encapsulates a Task with start date/time and end date/time.

Masunori committed:
Add three new methods:
- addToDo: add a to-do type task to the task list
- addDeadline: add a deadline type task to the task list
- addEvent: add an event type task to the task list
Add three new commands: todo, deadline, event
Remove a method: addToList.

Masunori committed:
Stop tracking test files in text-ui-test and now ignore all executables (.class files)

Masunori committed:
Add error handling

Masunori committed:
Fix minor string formatting for error handling in addEvent.

Masunori committed:
Untrack all files in test-ui-test

Masunori committed:
Untrack all files in test-ui-test

Masunori committed:
Untrack all files in test-ui-test

Masunori committed:
Untrack all files in test-ui-test

Masunori committed:
Add deleteTask method to delete a task. The chatbot now supports the 'delete' command.

Masunori committed:
Task class: add toNoteForm method for logging

Masunori committed:
Mirai.java: add log-to-file feature

Masunori committed:
tasks: move Task and subclasses to package

Masunori committed:
Mirai.java: handle invalid index for unmark, mark, delete

Masunori committed:
Split Mirai class into UI, storage, parser, task list

Masunori committed:
Refactor all date-time strings to LocalDateTime

Masunori committed:
Rearrange classes into ui, tasks and utility packages

Masunori committed:
Rearrange classes into ui, tasks and utility packages

Masunori committed:
Merge branch 'add-gradle-support'

Masunori committed:
Add Gradle support

Masunori committed:
Add JUnit tests fo Task and TaskList

Masunori committed:
Create a .jar file for the chatbot application

Masunori committed:
Add 'find' command to find all tasks based on keyword

Masunori committed:
Add JavaDoc comments to all classes and methods

Masunori committed:
Fix all code to comply standard

Masunori committed:
Merge branch 'branch-A-CodingStandard'

Masunori committed:
Merge branch 'branch-Level-9'

Masunori committed:
Fix JavaDoc typo

lauhengyi commented on src/main/java/mirai/tasks/Deadline.java:
I like the cleanliness of all the Task code (And its inheritors)

lauhengyi commented on src/main/java/mirai/utility/Storage.java:
I like how well the methods and files were named!

lauhengyi commented on src/main/java/mirai/utility/Ui.java:
Should greeting and goodbye be a class variable? Considering that they are only used in their respective methods (printGreeting and printGoodbye)?

lauhengyi commented on src/test/java/mirai/tasks/TaskTest.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Test
+    public void toDoToString_readBookMarked_success() {
```
I like the naming of your test cases

lauhengyi reviewed (commented):
Overall, I think the code looks clean and the naming looks appropriate. Great job!

Masunori committed:
TaskListTest: rearrange import statements to follow style

Masunori committed:
Replace CLI with GUI

The CLI is not intuitive for people unfamiliar with CLI

Let's implement a GUI using JavaFX and store them in mirai.ui package,
which will replace the original Ui class in the CLI version.
Refactor Command functional interface to return a message instead of a
flag to end or continue the conversation. Update methods in Parser
implementing this Command interface.

This is because the GUI does not need a loop to maintain the
conversation, unlike the CLI that contains a while loop under the hood
and needs a boolean flag to continue or terminate the loop. Instead
of the UI class printing the message to the terminal, a Command
directly returns the String message to feed into the DialogBox.

Masunori committed:
Command.java: modify execute() to take in varargs

Masunori committed:
Delete Ui.java

nlqhung130403 commented on src/main/java/mirai/ui/Mirai.java:
```diff
@@ -0,0 +1,40 @@
+ */
+public class Mirai {
+    private final Storage storage;
+    private TaskList tasks;
```
I like how you use plural form for collection object

nlqhung130403 commented on src/main/java/mirai/utility/Parser.java:
```diff
@@ -0,0 +1,386 @@
+/**
+ * The Parser class encapsulates a parser to interpret the user's command lines.
+ */
+public class Parser {
```
all the method names follow coding standard well. Great job!

nlqhung130403 commented on src/main/java/mirai/ui/Mirai.java:
```diff
@@ -0,0 +1,40 @@
+/**
+ * The Mirai class encapsulates a chatbot named Mirai.
+ */
+public class Mirai {
```
Overall, very nice design for your model.

nlqhung130403 commented on src/main/java/mirai/tasks/Event.java:
```diff
@@ -0,0 +1,42 @@
+
+    @Override
+    public String toNoteForm() {
+        return String.format("E | %d | %s | %s | %s",
```
Wow! Actually very concise and precise code

Masunori committed:
Fix event parsing to use correct dates from text file

Masunori committed:
Add assertions to TaskList

Operations on TaskList such as delete, mark or unmark do not
handle when the task index is not within the list range.

Even though the programmers have implemented pre-conditional
checks, the assertions will help detect errors if mistakes
can slip through.

Let's add assertions to delete, mark, unmark and getTask.

We use Java assert instead of JUnit assert for its
simplicity and less overhead (questionable).

Masunori committed:
Add assertions to TaskList

Operations on TaskList such as delete, mark or unmark do not
handle when the task index is not within the list range.

Even though the programmers have implemented pre-conditional
checks, the assertions will help detect errors if mistakes
can slip through.

Let's add assertions to delete, mark, unmark and getTask.

We use Java assert instead of JUnit assert for its
simplicity and less overhead (questionable).

Masunori committed:
Refactor Parser.java

The Parser.java class contains two methods addDeadline and
addEvent, which are currently too long and contain low-level
statements bundled together. This makes the code harder to
understand.

Breaking these two methods down based on functional groups will
allow better code readability and modularity.

The Parser's listAllSupportedCommands only lists commands without
any other information.

Listing the usage for each command will be more user-friendly, as the
user will know what to do.

Let's,
* break down the methods into three phases: parsing arguments,
parsing date-time, and generating task. Parsing arguments and
parsing date-time have been delegated to smaller methods that handle
the parses.
* implement a map from commands to message, where
listAllSupportedCommands can read from.

Masunori committed:
Create a dummy commit for A-Streams

Masunori committed:
Merge pull request #1 from Masunori/branch-A-Assertions

Branch A-Assertions

Masunori committed:
Merge branch 'master' into branch-A-CodeQuality

Masunori committed:
Merge pull request #4 from Masunori/branch-A-CodeQuality

Branch A-CodeQuality

Masunori committed:
Merge branch 'master' into branch-A-Streams

Masunori committed:
Merge pull request #5 from Masunori/branch-A-Streams

Branch a streams

Masunori committed:
Add .yml to set up GitHub Actions

Masunori committed:
Add .yml to set up GitHub Actions

Masunori committed:
Support flexible search

The 'find' command strictly checks for string matching to filter
out relevant tasks.

A more flexible search that instead, sort all events in order or
relevance will still display relevant results if the user forgets
the correct task description or makes a typo.

Let's
* implement a flexible search algorithm involving edit distance
* implement a new command called 'flexfind' in the Parser class
that uses this edit distance algorithm

The edit distance is used because of its simplicity to implement,
yet highly effective for simple matching.

Refer to the specific implementation at Task::computeCloseness
method.

Masunori committed:
Fix code to adhere to checkstyle

Masunori committed:
Fix code to adhere to checkstyle

Masunori committed:
Fix gradle.yml to allow run test.sh

Masunori committed:
Merge pull request #8 from Masunori/branch-BCD-Extension

Branch bcd extension

Masunori committed:
Remove I/O redirection tests for gradle.yml

Masunori committed:
Improve GUI and flexfind algorithm

The GUI is generic.

Choosing another colour theme for the GUI will make it less generic
and show that there is a certain theme associated with the chatbot.

Let's add styling to the JavaFX components.

CSS is used for styling because it concentrates all styling into
one file and style changes will be easier. Blue theme is used
because the imagined avatar of Mirai is a girl with blue hair and
eyes.

Next, the flexfind algorithm has a flaw. For each keyword, it
performs edit distance calculation and return the distance from
the closest word. For n keywords, the closeness is the sum of
n edit distances. However, when flexfinding with 'r', 'wake up'
ranks higher than 'read book' because 'r' is closer to 'up' than
'read'.

The improved algorithm must also account for inclusion, rather than
just edit distance to the closest word.

Let's add another logic layer to the algorithm, where if a description
word contains a key word, edit distance is straight up 0.

Masunori committed:
Remove redundant files

Masunori committed:
Fix styling issues

Masunori committed:
Merge pull request #11 from Masunori/branch-A-BetterGui

Branch a better gui

Masunori committed:
Fix styling issues

Masunori committed:
Merge pull request #12 from Masunori/branch-A-BetterGui

Fix styling issues

Masunori committed:
Update README.md

Masunori committed:
Add application title and correct Message.java's grammar

Masunori committed:
Merge branch 'master' of https://github.com/Masunori/ip

Masunori committed:
Update README.md

Masunori committed:
Update README.md

Masunori committed:
Update README.md

Masunori committed:
Update README.md

Masunori committed:
Update README.md

Masunori committed:
Update README.md

Masunori committed:
Update README.md

Masunori committed:
Fix README.md
